src/coinsviewcacheasync.h

```diff
@@ -0,0 +1,241 @@
+            if (auto coin{GetCoinWithoutMutating(outpoint)}) {
+                ret->second.coin = std::move(*coin);
+            } else {
+                cacheCoins.erase(ret);
+                return cacheCoins.end();
+            }
+        }
+
+        cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage();
+        return ret;
+    }
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
+
+    //! Stop all worker threads.
+    void StopFetching() noexcept
+    {
+        if (m_inputs.size() == 0) return;
+        // Skip fetching the rest of the inputs by moving the head to the end.
+        m_input_head.store(m_inputs.size(), std::memory_order_relaxed);
+        // Wait for all threads to stop.
+        m_barrier.arrive_and_wait();
+        m_inputs.clear();
+    }
+
+public:
+    //! Start fetching all block inputs in parallel.
+    void StartFetching(const CBlock& block) noexcept
+    {
+        // Loop through the inputs of the block and set them in the queue. Also construct the set of txids to filter.
+        for (const auto& tx : block.vtx | std::views::drop(1)) {
+            for (const auto& input : tx->vin) m_inputs.emplace_back(input.prevout);
+            m_txids.emplace_back(tx->GetHash().ToUint256().GetUint64(0));
+        }
+        // Don't start threads if there's nothing to fetch.
+        if (m_inputs.size() == 0) return;
+        // Sort txids so we can do binary search lookups.
+        std::ranges::sort(m_txids);
+        // Start workers by entering the barrier.
+        m_barrier.arrive_and_wait();
+    }
+
+    //! Stop fetching and reset state. Must be called before block is destroyed.
+    void Reset() noexcept
+    {
+        StopFetching();
+        m_input_head.store(0, std::memory_order_relaxed);
+        m_input_tail = 0;
+        m_txids.clear();
+        cacheCoins.clear();
+        cachedCoinsUsage = 0;
+        hashBlock = uint256::ZERO;
+    }
+
+    bool Flush() override
+    {
+        // We need to stop workers from accessing base before we mutate it.
+        StopFetching();
+        auto cursor{CoinsViewCacheCursor(m_sentinel, cacheCoins, /*will_erase=*/true)};
+        const auto ret{base->BatchWrite(cursor, hashBlock)};
+        Reset();
+        return ret;
+    }
+
+    explicit CoinsViewCacheAsync(CCoinsViewCache& cache, const CCoinsView& db, bool deterministic = false) noexcept
+        : CCoinsViewCache{&cache, deterministic}, m_db{db}, m_barrier{WORKER_THREADS + 1}
+    {
+        for (const auto n : std::views::iota(0U, WORKER_THREADS)) {
+            m_worker_threads.emplace_back([this, n] {
+                util::ThreadRename(strprintf("inputfetcher.%i", n));
+                while (true) {
+                    m_barrier.arrive_and_wait();
+                    while (ProcessInputInBackground()) {}
+                    if (m_inputs.size() == 0) return;
```

Comment:
I understand that `size() > 0` may be more descriptive than `!empty()`, but there's a dedicated method for this case (there are a few more cases):
```suggestion
                    if (m_inputs.empty()) return;
```